src/spec-common/injectHeadless.ts

```diff
@@ -548,38 +552,86 @@ export function createFileCommand(location: string) {
 	return `test ! -f '${location}' && set -o noclobber && mkdir -p '${path.posix.dirname(location)}' && { > '${location}' ; } 2> /dev/null`;
 }
 
-export async function runRemoteCommand(params: { output: Log; onDidInput?: Event<string> }, { remotePtyExec }: { remotePtyExec: PtyExecFunction }, cmd: string[], cwd?: string, options: { remoteEnv?: NodeJS.ProcessEnv; stdin?: Buffer | fs.ReadStream; silent?: boolean; print?: 'off' | 'continuous' | 'end'; resolveOn?: RegExp } = {}) {
-	const print = options.print || (options.silent ? 'off' : 'end');
-	const p = await remotePtyExec({
-		env: options.remoteEnv,
-		cwd,
-		cmd: cmd[0],
-		args: cmd.slice(1),
-		output: options.silent ? nullLog : params.output,
-	});
+export async function runRemoteCommand(params: { output: Log; onDidInput?: Event<string>; stdin?: NodeJS.ReadStream; stdout?: NodeJS.WriteStream; stderr?: NodeJS.WriteStream }, { remoteExec, remotePtyExec }: ContainerProperties, cmd: string[], cwd?: string, options: { remoteEnv?: NodeJS.ProcessEnv; pty?: boolean; print?: 'off' | 'continuous' | 'end' } = {}) {
+	const print = options.print || 'end';
+	let sub: Disposable | undefined;
+	let pp: Exec | PtyExec;
 	let cmdOutput = '';
-	let doResolveEarly: () => void;
-	const resolveEarly = new Promise<void>(resolve => {
-		doResolveEarly = resolve;
-	});
-	p.onData(chunk => {
-		cmdOutput += chunk;
-		if (print === 'continuous') {
-			params.output.raw(chunk);
+	if (options.pty) {
+		const p = pp = await remotePtyExec({
+			env: options.remoteEnv,
+			cwd,
+			cmd: cmd[0],
+			args: cmd.slice(1),
+			output: params.output,
+		});
+		p.onData(chunk => {
+			cmdOutput += chunk;
+			if (print === 'continuous') {
+				if (params.stdout) {
+					params.stdout.write(chunk);
+				} else {
+					params.output.raw(chunk);
+				}
+			}
+		});
+		if (params.onDidInput) {
+			params.onDidInput(data => p.write(data));
+		} else if (params.stdin) {
+			const listener = (data: Buffer): void => p.write(data.toString());
+			const stdin = params.stdin;
+			if (stdin.isTTY) {
+				stdin.setRawMode(true);
+			}
+			stdin.on('data', listener);
+			sub = { dispose: () => stdin.off('data', listener) };
 		}
-		if (options.resolveOn && options.resolveOn.exec(cmdOutput)) {
-			doResolveEarly();
+	} else {
+		const p = pp = await remoteExec({
```

Comment:
why do you need to assign this twice?

Reply:
`pp` has a wider type than `p`. `p` is used inside the else with specific properties while `pp` is also assigned in the "then" block and used outside the if-then-else with less specific properties (the `exit` property).